src/AWSS3.jl

```diff
@@ -800,6 +779,111 @@ function s3_sign_url(aws::AWSConfig, bucket, path, seconds=3600;
+                                        "X-Amz-Algorithm" => scheme * "-" * algorithm,
+                                        "X-Amz-Credential" => aws[:creds].access_key_id * "/" * scope,
+                                        "X-Amz-Date" => datetime_stamp,
+                                        "X-Amz-Security-Token" => aws[:creds].token,
+                                        "X-Amz-SignedHeaders" => canonical_header_names)
+
+    if !isempty(aws[:creds].token)
+        query["X-Amz-Security-Token"] = aws[:creds].token
+    end
+
+    sort!(query; by = name -> lowercase(name))
+
+    canonical_headers = join(map(header -> lowercase(header.first) * ":" * lowercase(header.second) * "\n", collect(headers)))
+
+    canonical_request = verb * "\n" *
+                        path * "\n" *
+                        HTTP.escapeuri(query) * "\n" *
+                        canonical_headers * "\n" *
+                        canonical_header_names * "\n" *
+                        "UNSIGNED-PAYLOAD"
+
+    string_to_sign = scheme * "-" * algorithm * "\n" *
+                     datetime_stamp * "\n" *
+                     scope * "\n" *
+                     bytes2hex(digest(MD_SHA256, canonical_request))
+
+    key_secret = scheme * aws[:creds].secret_key
+    key_date = digest(MD_SHA256, date_stamp, key_secret)
+    key_region = digest(MD_SHA256, aws[:region], key_date)
+    key_service = digest(MD_SHA256, service, key_region)
+    key_signing = digest(MD_SHA256, terminator, key_service)
+    signature = digest(MD_SHA256, string_to_sign, key_signing)
+
+    query["X-Amz-Signature"] = bytes2hex(signature)
+
+    return string(protocol, "://", host, path, "?", HTTP.escapeuri(query))
+end
+
+
+"""
+    s3_sign_url([::AWSConfig], bucket, path, [seconds=3600];
+                [verb="GET"], [content_type="application/octet-stream"], [signature_version="v4"])
+
+Create a
+[pre-signed url](http://docs.aws.amazon.com/AmazonS3/latest/dev/ShareObjectPreSignedURL.html) for `bucket` and `path` (expires after for `seconds`).
+
+To create an upload URL use `verb="PUT"` and set `content_type` to match
+the type used in the `Content-Type` header of the PUT request.
+
+For compatibility, the signature version 2 signing process can be used by setting `signature_version="v2"`
+but it is [recommended](https://docs.aws.amazon.com/general/latest/gr/signature-version-2.html) that the default version 4 is used.
+
+```
+url = s3_sign_url("my_bucket", "my_file.txt"; verb="PUT")
+Requests.put(URI(url), "Hello!")
+```
+```
+url = s3_sign_url("my_bucket", "my_file.txt";
+                  verb="PUT", content_type="text/plain")
+
+Requests.put(URI(url), "Hello!";
+             headers=Dict("Content-Type" => "text/plain"))
+```
+"""
+function s3_sign_url(aws::AWSConfig, bucket, path, seconds=3600;
+                     verb="GET", content_type="application/octet-stream",
+                     protocol="http", signature_version="v4")
```

Comment:
This is probably getting way too into it, but we could do what is below. However, it's probably overkill. 🤷 

```
export SIGNATURE_VERSION
@enum SIGNATURE_VERSION v2=0 v4=1

function s3_sign_url(
    ::Type{Val{v2}}, aws::AWSConfig, bucket, path, seconds=3600; 
    verb="GET", content_type="application/octet-stream", protocol="http",
)
    # v2 signing code here
end

function s3_sign_url(
    ::Type{Val{v4}}, aws::AWSConfig, bucket, path, seconds=3600; 
    verb="GET", content_type="application/octet-stream", protocol="http",
)
    # v4 signing code here
end
```

Reply:
I have no opinion either way. I'd imagine we want to have a consistent style/format across all packages under JuliaCloud?